### PR TITLE
ESQL: Fix error on sorting unsortable geo_point and cartesian_point

### LINKED ISSUE #106007: ESQL: TopN on geo_point or cartesian_point field types fails incorrectly
### Description

`FROM airports | limit 10 | sort city_location desc nulls last` fails with 

```
                "type": "illegal_state_exception",
                "reason": "Cannot find encoder for BytesRef value",
                "stack_trace": "org.elasticsearch.ElasticsearchException$1: Cannot find encoder for BytesRef value
```

should instead fail with `can't sort on geo_point or point field without using specific sorting feature, like geo_distance` I believe.

Same thing happens for a `cartesian_point`: `ROW wkt = [\"POINT(4297.11 -1475.53)\", \"POINT(7580.93 2272.77)\"] | MV_EXPAND wkt | EVAL pt = TO_CARTESIANPOINT(wkt) | limit 5 | sort pt`

### PR DESCRIPTION
**Pull request by craigtaverner**
Without a LIMIT the correct error worked, but with LIMIT it did not. This fix mimics the same error with LIMIT and adds tests for all three scenarios:

* Without limit
* With Limit
* From row with limit

Fixes #106007

### CODE REVIEW

**Review by astefan — CHANGES_REQUESTED** [review]
I don't believe this is the best place for such a check. Have a look at the `Verifier` for this instead.

Also, it's not clear to me (due to lack of geo knowledge), `geo_distance` is an ES specific query. Having this appear as an ESQL error message is not right. When I created the issue I wrongly assumed the error `can't sort on geo_point or point field without using specific sorting feature, like geo_distance` comes from ESQL, but it's ES itself.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java [hunk 1]
```diff
@@ -381,6 +384,19 @@ private static Failure validateUnsignedLongNegation(Neg neg) {
         return null;
     }
 
+    /**
+     * Makes sure that spatial types do not appear in sorting contexts.
+     * This check is a simplification of a similar check in the SQL verifier.
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:389]: This statement is not needed imo.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java [hunk 2]
```diff
@@ -381,6 +384,19 @@ private static Failure validateUnsignedLongNegation(Neg neg) {
...
+     */
+    private static void checkForSortOnSpatialTypes(LogicalPlan p, Set<Failure> localFailures) {
+        p.forEachDown(OrderBy.class, o -> o.order().forEach(agg -> agg.forEachUp(FieldAttribute.class, fa -> {
+            DataType dataType = fa.field().getDataType();
+            if (EsqlDataTypes.isSpatial(dataType)) {
+                localFailures.add(fail(fa, "can't sort on " + dataType.typeName()));
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:395]: Use the longer version, please. `cannot`

**Review by astefan — COMMENTED** [review]
Please, add unit tests for a wider set of queries (sort, sort + limit, any other place where this is not supported) in `VerifierTests`. Also, use `row` commands in your unit tests.

**Comment by astefan** [comment]
And we need a statement in Documentation -> Limitations pages, as well.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java (hunk 1 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java]: Removed

**Review by astefan — COMMENTED** [review]
Looks good, but add some tests with the `from` command, as well, to VerifierTests. Sorry, I should have been more explicit last time I mentioned VerifierTests.

**Comment by craigtaverner** [comment]
> Looks good, but add some tests with the `from` command, as well, to VerifierTests. Sorry, I should have been more explicit last time I mentioned VerifierTests.

OK. Added those too.

**Review by astefan — APPROVED** [review]
LGTM. Thanks

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java [hunk 3]
```diff
@@ -381,6 +384,18 @@ private static Failure validateUnsignedLongNegation(Neg neg) {
         return null;
     }
 
+    /**
+     * Makes sure that spatial types do not appear in sorting contexts.
+     */
+    private static void checkForSortOnSpatialTypes(LogicalPlan p, Set<Failure> localFailures) {
+        p.forEachDown(OrderBy.class, o -> o.order().forEach(agg -> agg.forEachUp(Attribute.class, fa -> {
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:391]: The outer loop already does forDown, so no need for you to do it. Also Order is an expression so you don't have to unroll the list; instead do:
```
if (p instanceof OrderBy ob) {
    ob.forEachExpressionOnly(Attribute, fa -> ...)
}
```

Similar to what the other methods are doing.

**Review by costin — APPROVED** [review]
Left some comments otherwise LGTM

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java (hunk 3 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java]: I had copied the similar code from the SQL module, so did not think about the duplicate forEachDown, but I've removed it now. The code has more lines, but is more readable and no longer has the duplicated forEachDown.

**Pull request merged**